### PR TITLE
JDK-8307331: Correctly update line maps when class redefine rewrites bytecodes

### PR DESCRIPTION
**Pull request by adinn**
This small change ensures that repeated bytecode rewrites necessitated by class pool index updates are applied cumulatively when updating the method line number table. The current code applies each change to the original table which means only the last one is applied (and even then with the wrong adjustment).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307331](https://bugs.openjdk.org/browse/JDK-8307331): Correctly update line maps when class redefine rewrites bytecodes

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13795/head:pull/13795` \
`$ git checkout pull/13795`

Update a local copy of the PR: \
`$ git checkout pull/13795` \
`$ git pull https://git.openjdk.org/jdk.git pull/13795/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13795`

View PR using the GUI difftool: \
`$ git pr show -t 13795`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13795.diff">https://git.openjdk.org/jdk/pull/13795.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13795#issuecomment-1534403851)

### CODE REVIEW

**Comment by adinn** [comment]
This fix has been tested manually using the test case linked from the JIRA. An automated test might be possible but is difficult to perform as it would involve hot swapping code and checking line number adjustments via JDWP. Advice from the serviceability team on how to implement such a test would be welcome.

**Comment by adinn** [comment]
@coleenp @plummercj Any chance of feedback or a review for this patch?

**Comment by DanHeidinga** [comment]
@adinn Looking at the closely related code, is the same problem present for `adjust_exception_table` & `adjust_local_var_table`?  Both appear to always reach for the original value from the method, though unlike the line number table, there's no member variable cached in the Relocator for either of them.

**Comment by plummercj** [comment]
@adinn This is not code I'm at all familiar with. Perhaps @sspitsyn can help.

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thank you for taking care about it!
Thanks,
Serguei

**Comment by adinn** [comment]
> @adinn Looking at the closely related code, is the same problem present for `adjust_exception_table` & `adjust_local_var_table`? Both appear to always reach for the original value from the method, though unlike the line number table, there's no member variable cached in the Relocator for either of them.

@DanHeidinga  I also thought that at first -- but it turns out the answer is no. Those other cases are different because the relocator directly updates the relevant offsets in the data located at the end of the ConstMethod. The line number table needs to be uncompressed and recompressed at each update, possibly ending up with a different size compressed array. So, the changes need to accumulate in a succession of compressed arrays held in the relocator the last of which gets recombined with the new version of the method  when the method clone eventually happens.

**Comment by adinn** [comment]
@sspitsyn Thanks for the review.

**Comment by dholmes-ora** [comment]
@adinn Please wait for two reviews for hotspot changes unless designated as trivial. Thanks.

**Comment by adinn** [comment]
@dholmes-ora Apologies for jumping the gun. Do I need to obtain a second, retrospective review?

**Comment by sspitsyn** [comment]
> Do I need to obtain a second, retrospective review?

Yes, Serviceability changes need two reviews unless it is not as trivial.

**Comment by adinn** [comment]
@sspitsyn @dholmes-ora Could you suggest a suitable reviewer?

**Comment by dholmes-ora** [comment]
I can't specifically suggest anyone but will try to find someone. I looked but don't know the code.

**Comment by adinn** [comment]
@dholmes-ora Thanks. Perhaps @coleenp might be able to review it?

**Review by tstuefe — COMMENTED** [review]
This looks fine to me. Thank you for the well written description text in the JBS issue.

I hesitated a bit at the use of TraceRelocator instead of our usual LogStream/LogTarget combination with UL, but I see that this is pre-existing usage, so its fine.

Another preexisting issue I noted is that CompressedWriteStream manages its internal buffer via RA, and grows inside RA too. That is suboptimal since we usually cannot reclaim the old buffer on resize since its rarely the top arena allocation. Oh, we also just allocate, we never even try resizing... (CompressedWriteStream::grow()).

**Review by coleenp — COMMENTED** [review]
Sorry I didn't see this in my mail until just now.  Wow, good analysis and makes sense.  Late second review.

**Pull request closed without merging**
